src/main/kotlin/com/example/ajaxproject/pbp/NotificationAnnotationBeanPostProcessor.kt

```diff
@@ -19,74 +23,73 @@ import kotlin.reflect.full.memberFunctions
-        return beans[beanName]?.let { beanClass ->
+        return if (serviceBeans.containsKey(beanName)) {
             Proxy.newProxyInstance(
-                beanClass.java.classLoader,
-                beanClass.java.interfaces,
-                NotificationInvocationHandler(bean, groupChatRoomRepository, emailSenderService, beanClass)
+                bean::class.java.classLoader,
+                bean::class.java.interfaces,
+                NotificationInvocationHandler(bean, beanFactory)
             )
-        } ?: bean
+        } else {
+            bean
+        }
     }
 }
 
-@Suppress("SpreadOperator")
 class NotificationInvocationHandler(
     private val bean: Any,
-    private val groupChatRoomRepository: GroupChatRoomRepository,
-    private val emailSenderService: EmailSenderService,
-    private val originalBean: KClass<*>
+    private val beanFactory: BeanFactory
 ) : InvocationHandler {
     override fun invoke(proxy: Any, method: Method, args: Array<out Any>?): Any? {
+        val result = method.invoke(bean, *args.orEmpty())
 
-        val methodParams = args ?: emptyArray()
-        val hasNotificationAnnotation = originalBean.memberFunctions.any { beanMethod ->
-            beanMethod.name == method.name &&
-                    beanMethod.annotations.any { it is Notification }
+        if (args?.any { it is ChatDTO } == true) {
+            createNotification(args.find { it is ChatDTO } as ChatDTO)
         }
-        if (hasNotificationAnnotation) {
-            val userRequest = args?.find { it is GroupChatDTO } as? GroupChatDTO
-            if (userRequest != null) {
-                createNotification(userRequest)
-            }
-        }
-        return method.invoke(bean, *methodParams)
+        return result
     }
 
-    private fun createNotification(groupChatDTO: GroupChatDTO) {
-
-        val chatName = groupChatRoomRepository.findChatRoom(groupChatDTO.chatId).chatName
+    private fun createNotification(chatDTO: ChatDTO) = runBlocking {
+        val groupChatRoomRepository = beanFactory.getBean(GroupChatRoomRepository::class.java)
+        val chatName = groupChatRoomRepository.findChatRoom(chatDTO.chatId).chatName
+        val userList = groupChatRoomRepository.findChatRoom(chatDTO.chatId).chatMembers
+
+        val emailSenderService = beanFactory.getBean(EmailSenderService::class.java)
+
+        val deferredEmails = userList.map { user ->
+            CoroutineScope(Dispatchers.Default).async {
+                val emailDTO = EmailDTO(
+                    from = "ora.romaniuk@gmail.com",
+                    to = user.email,
+                    subject = "Testing post bean processor",
+                    body = "New message in chat $chatName"
+                )
+                emailSenderService.send(emailDTO)
+            }
+        }
 
-        val userList = groupChatRoomRepository.findChatRoom(groupChatDTO.chatId).chatMembers
+        deferredEmails.forEach { it.await() }
```

Comment:
Optional suggestion

```kotlin
        userList.asSequesnce()
            .map { user -> buildEmail(user.email) }
            .map { email -> 
                CoroutineScope(Dispatchers.Default).async {
                    emailSenderService.send(emailDTO) 
                }
            }
            .toList()
            .forEach { it.await() }

private fun buildEmail(email: String): EmailDTO = EmailDTO(
    from = "ora.romaniuk@gmail.com",
    to = email,
    subject = "Testing post bean processor",
    body = "New message in chat $chatName"
)
```

Reply:
Did in the same way

Reply:
Also maybe the email from which I send it should be hidden somewhere?

Reply:
it's better to move email(send from) to `application.properties`